Runtime/Native/Windows/NativeClient.cs

```diff
@@ -0,0 +1,438 @@
+                return;
+            }
+
+            foreach (var attribute in clientAttributes)
+            {
+                AddNativeAttribute(attribute.Key, attribute.Value);
+            }
+
+            // add exception type to crashes handled by crashpad - all exception handled by crashpad 
+            // by default we setting this option here, to set error.type when unexpected crash happen (so attribute will present)
+            // otherwise in other methods - ANR detection, OOM handler, we're overriding it and setting it back to "crash"
+
+            // warning 
+            // don't add attributes that can change over the time to initialization method attributes. Crashpad will prevent from 
+            // overriding them on game runtime. ANRs/OOMs methods can override error.type attribute, so we shouldn't pass error.type 
+            // attribute via attributes parameters.
+            AddNativeAttribute("error.type", "Crash");
+        }
+
+        public void Disable()
+        {
+            if (_anrThread != null)
+            {
+                _stopAnr = true;
+            }
+            return;
+        }
+
+        public void GetAttributes(IDictionary<string, string> attributes)
+        {
+            return;
+        }
+
+        public void HandleAnr(string gameObjectName = "", string callbackName = "")
+        {
+            if (!_captureNativeCrashes || !_configuration.HandleANR)
+            {
+                return;
+            }
+
+            bool reported = false;
+            var mainThreadId = Thread.CurrentThread.ManagedThreadId;
+            _anrThread = new Thread(() =>
+            {
+                float lastUpdatedCache = 0;
+                while (_anrThread.IsAlive && _stopAnr == false)
+                {
+                    if (!_preventAnr)
+                    {
+                        if (lastUpdatedCache == 0)
+                        {
+                            lastUpdatedCache = _lastUpdateTime;
+                        }
+                        else if (lastUpdatedCache == _lastUpdateTime)
+                        {
+                            if (!reported)
+                            {
+
+                                reported = true;
+                                if (AndroidJNI.AttachCurrentThread() == 0)
+                                {
+                                    // set temporary attribute to "Hang"
+                                    AddNativeAttribute("error.type", "Hang");
+
+                                    NativeReport("ANRException: Blocked thread detected.", true);
+                                    // update error.type attribute in case when crash happen 
+                                    AddNativeAttribute("error.type", "Crash");
+                                }
+                            }
+                        }
+                        else
+                        {
+                            reported = false;
```

Comment:
Won't this logic still just keep sending infinite ANR reports while `lastUpdatedCache == _lastUpdateTime`? Every iteration we will set `reported = true`, then next iteration we will again set `reported = false` and send another ANR report.

Reply:
we won't enter this else statement when we're in the middle of ANR (lastUpdatedCache == _lastUpdateTime)

Reply:
But ANR thread is a separate background thread running a while loop, it is not tied to game update logic. It is possible that `lastUpdatedCache == _lastUpdateTime` while `_lastUpdateTime` doesn't change for a long time, causing us iterate this loop for an indefinite number of times while game engine is not responsive.

Reply:
> But ANR thread is a separate background thread running a while loop, it is not tied to game update logic. 
It's not true. Background thread shares a reference to _lastUpdateTime with main thread. Every time when main thread updates _lastUpdateTime, the background thread is able to detect it and then set it to `false`.

```
 It is possible that lastUpdatedCache == _lastUpdateTime while _lastUpdateTime doesn't change for a long time, causing us iterate this loop for an indefinite number of times while game engine is not responsive.
```
Based on what I see and how it works it's impossible for me - I don't see a way how it's possible.

Reply:
Look at this loop, it will iterate as fast as possible, this loop is not bound to the Unity game `Update` method:
`                while (_anrThread.IsAlive && _stopAnr == false)`

This means that even if `lastUpdatedCache == _lastUpdateTime` only for one Unity update frame (8 - 16 ms depending on frame rate), the while loop will still go into this path many times, potentially sending multiple reports:
`                        else if (lastUpdatedCache == _lastUpdateTime)`

Does that make sense? Because while loop will try to execute as fast as possible it can happen many times even within one frame